WordPress/src/main/java/org/wordpress/android/ui/stories/StoryComposerViewModel.kt

```diff
@@ -0,0 +1,97 @@
+
+class StoryComposerViewModel @Inject constructor(
+    private val systemNotificationsTracker: SystemNotificationsTracker,
+    private val saveInitialPostUseCase: SaveInitialPostUseCase,
+    private val analyticsUtilsWrapper: AnalyticsUtilsWrapper,
+    private val dispatcher: Dispatcher
+) :
+        ViewModel() {
+    private lateinit var editPostRepository: EditPostRepository
+    private lateinit var site: SiteModel
+    private lateinit var postEditorAnalyticsSession: PostEditorAnalyticsSession
+
+    fun start(
+        site: SiteModel,
+        editPostRepository: EditPostRepository,
+        postId: LocalId,
+        intent: Intent,
+        postEditorAnalyticsSession: PostEditorAnalyticsSession?,
+        notificationType: NotificationType?
+    ) {
+        this.editPostRepository = editPostRepository
+        this.site = site
+
+        if (postId.value == 0) {
+            // Create a new post
+            saveInitialPostUseCase.saveInitialPost(editPostRepository, site)
+            // Bump post created analytics only once, first time the editor is opened
+            analyticsUtilsWrapper.trackEditorCreatedPost(
+                    intent.action,
+                    intent,
+                    site,
+                    editPostRepository.getPost()
+            )
+        } else {
+            editPostRepository.loadPostByLocalPostId(postId.value)
+        }
+
+        setupPostEditorAnalyticsSession(postEditorAnalyticsSession)
+
+        notificationType?.let {
+            systemNotificationsTracker.trackTappedNotification(it)
+        }
+    }
+
+    private fun setupPostEditorAnalyticsSession(postEditorAnalyticsSession: PostEditorAnalyticsSession?) {
+        this.postEditorAnalyticsSession = postEditorAnalyticsSession ?: createPostEditorAnalyticsSessionTracker(
+                editPostRepository.getPost(),
+                site
+        )
+        this.postEditorAnalyticsSession.start(null)
+    }
+
+    private fun createPostEditorAnalyticsSessionTracker(
+        post: PostImmutableModel?,
+        site: SiteModel?
+    ): PostEditorAnalyticsSession {
+        return PostEditorAnalyticsSession.getNewPostEditorAnalyticsSession(
+                PostEditorAnalyticsSession.Editor.WP_STORIES_CREATOR,
+                post, site, true
+        )
+    }
+
+    fun writeToBundle(outState: Bundle) {
+        outState.putSerializable(WordPress.SITE, site)
+        outState.putInt(StoryComposerActivity.STATE_KEY_POST_LOCAL_ID, editPostRepository.id)
+        outState.putSerializable(StoryComposerActivity.STATE_KEY_EDITOR_SESSION_DATA, postEditorAnalyticsSession)
+    }
+
+    fun onStoryDiscarded() {
+        // delete empty post from database
+        dispatcher.dispatch(PostActionBuilder.newRemovePostAction(editPostRepository.getEditablePost()))
+        postEditorAnalyticsSession.setOutcome(CANCEL)
+    }
+
+    override fun onCleared() {
+        super.onCleared()
+        postEditorAnalyticsSession.end()
```

Comment:
I moved this from the `onDestroy` of the `StoryComposerActivity` I am just putting this comment here to verify that this will behave correctly with the `ViewModel`s lifecycle. I will investigate to verify.

Reply:
I checked this out and using `onCleared()` is fine because the lifecycle of the `ViewModel` is scoped to the component that it gets attached to. In this case, the `onCleared()` is directly associated with the activity's `onDestroy()`.

Reply:
Just a minor note: This change will change the behavior a bit. When you for example rotate your device "onDestroy" would have been called and session end tracked. But now, it gets tracked only when you leave the screen. I guess that's what we want anyway, but I thought I'd better mention it.

Reply:
Good notes there!  just leaving another note here as well 😅 : given the StoryComposerActivity is fixed to portrait for product design reasons, rotation in particular wouldn't have affected it  - of course the Activity would still be re-created on other configuration changes still 👍 so, good call moving it to the ViewModel's `onClear` so the session is kept until the user actually goes away from the Activity